Update page on smart contract composability (#6265)
ethereum/ethereum-org-website: src/content/developers/docs/smart-contracts/composability/index.md

## A brief introduction {#a-brief-introduction}

> Smart contracts are public on Ethereum and can be thought of as open APIs. You don't need to write your own smart contract to become a dapp developer, you just need to know how to interact with them. 
> 
> For example, you can use the existing smart contracts of [Uniswap](https://uniswap.exchange/swap), a decentralized exchange, to handle all the token swap logic in your app – you don't need to start from scratch. Check out some of their [v2](https://github.com/Uniswap/uniswap-v2-core/tree/master/contracts) and [v3](https://github.com/Uniswap/uniswap-v3-core/tree/main/contracts) contracts.

Reviewer: I don't think we should break onto a new line here. 

```suggestion
Smart contracts are public on Ethereum and can be thought of as open APIs. You don't need to write your own smart contract to become a dapp developer, you just need to know how to interact with them. For example, you can use the existing smart contracts of [Uniswap](https://uniswap.exchange/swap), a decentralized exchange, to handle all the token swap logic in your app – you don't need to start from scratch. Check out some of their [v2](https://github.com/Uniswap/uniswap-v2-core/tree/master/contracts) and [v3](https://github.com/Uniswap/uniswap-v3-core/tree/main/contracts) contracts.
```

## What is composability? {#what-is-composability}

> Composability is the idea that distinct components can be integrated or combined to create new systems or outputs. In the context of blockchain development, composability means existing pieces of software can be reused to build new applications.
> 
> A good way to understand composability is to think of composable elements as Lego blocks. Each Lego is designed to fit into another, such that you can build complex structures by combining different Lego blocks. 

Reviewer: This should be one paragraph. Tweaked the wording slightly but I love this :)

```suggestion
Composability is combing distinct components to create new systems or outputs. In software development, composability means developers can reuse existing software components to build new applications. A good way to understand composability is to think of composable elements as Lego blocks. Each Lego can be combined with another, allowing you to build complex structures by combining different Legos. 
```

In Ethereum, every smart contract is a Lego of sorts—you can use smart contracts from other projects as building blocks for your project. This means you don't have to spend time reinventing the wheel or building from scratch.

## How does composability work? {#how-does-composability-work}

> As we said in the beginning, Ethereum smart contracts are like public APIs. This means anyone can interact with them or integrate them into dapps for added functionality.
> 
> Smart contract composability generally works off three principles: modularity, autonomy, and discoverability: 

Reviewer: ```suggestion
Ethereum smart contracts are like public APIs, so anyone can interact with the contract or integrate them into dapps for added functionality. Smart contract composability generally works off three principles: modularity, autonomy, and discoverability:  
```

**1. Modularity**: This is the ability of individual components to perform a specific task. In Ethereum, every smart contract has specific use-case (as shown in the Uniswap example).

**2. Autonomy**: Composable components must be able to operate independently. Each smart contract in Ethereum is self-executing and can function without relying on other parts of the system. 

**3. Discoverability**: Developers cannot call external contracts or integrate software libraries into applications if the former are not publicly available. By design, smart contracts are open-source; anyone can call a smart contract or fork a codebase. 

## Benefits of composability {#benefits-of-composability}

> ### Shorter development cycle

Reviewer: Could you please add header-ids to all the H3s 🙏, like this:

```suggestion
### Shorter development cycle {#shorter-development-cycle}
```

> Composability reduces the work that developers have to do when creating [decentralized applications](https://ethereum.org/en/dapps/#what-are-dapps) (dapps). As Naval Ravikant [puts it](https://twitter.com/naval/status/1444366754650656770): "Open source means every problem has to be solved once."

Reviewer: - Let's assume readers of this page know what a dapp is
- Tip: when linking internally on ethereum.org you can use a relative path, (e.g. https://ethereum.org/en/developers/docs/smart-contracts/ becomes /developers/docs/smart-contracts/)

```suggestion
Composability reduces the work that developers have to do when creating [dapps](/dapps/#what-are-dapps). [As Naval Ravikant puts it:](https://twitter.com/naval/status/1444366754650656770) "Open source means every problem has to be solved once."
```

Author: Thanks for the tip on relative paths--I'm just learning about those now.

> Once a particular smart contract solves one problem, other developers can reuse it, so they don’t have to solve the same problem. In this way, developers can take existing software libraries and add extra functionality to create new dapps. 

Reviewer: ```suggestion
If there is a smart contract that solves one problem, other developers can reuse it, so they don’t have to solve the same problem. This way, developers can take existing software libraries and add extra functionality to create new dapps.
```

Reviewer: Is there any argument against this being one paragraph?

### Greater innovation

> Composability encourages innovation and experimentation. That’s because developers are free to reuse, modify, duplicate, or integrate open-source code to create desired results. 

Reviewer: ```suggestion
Composability encourages innovation and experimentation because developers are free to reuse, modify, duplicate, or integrate open-source code to create desired results. As a result, development teams spend less time on basic functionality and can allocate more time experimenting with new features.
```

> Also, development teams spend less time on basic functionality and can allocate more time to experimenting with new features. This explains why there are so many great projects in the Ethereum ecosystem today—developers can focus on improve on existing dapps and build better products. 

Reviewer: > This explains why there are so many great projects in the Ethereum ecosystem today—developers can focus on improve on existing dapps and build better products. 

We try to avoid subjective opinions where possible.

Author: Understood. I'll note it for future contributions.

> Also, development teams spend less time on basic functionality and can allocate more time to experimenting with new features. This explains why there are so many great projects in the Ethereum ecosystem today—developers can focus on improve on existing dapps and build better products. 

Reviewer: Merged into the above paragraph 

```suggestion
```

### Better user experience

> Interoperability between components of the Ethereum ecosystem improves the user experience. Users can access greater functionality when dapps integrate external smart contracts than in a fragmented ecosystem where applications cannot communicate.

Reviewer: This feels very vague to me. Better user experience because of more feature-rich dapps? I'd (weakly) disagree with this point.

Author: What I had in mind was how interactions between different smart contracts increases the functionality available to users. I think flash loans involving multiple dapps are a good example of this (AMMs also fit into this category). It'd be difficult to have a highly interoperable system like that if everyone walled off their apps.

Reviewer 2: Yeah, I see what you're getting at here @emmanuel-awosika , but I think it may help to clarify with at least one example.

Author: Hi @wackerow. Thanks for the feedback. I added the flash loan example to explain the point. Let me know if that works.

## Examples of composability in Ethereum {#composability-in-ethereum}

### Token swaps

> Say you have a dapp that requires to pay for transactions in ETH. You can add more payment options by integrating token swap logic from Uniswap contracts. This code will automatically convert the user’s token to ETH before the contract executes the called function.

Reviewer: ```suggestion
If you create a dapp that requires transactions to be paid in ETH, you can allow users to pay in other ERC-20 tokens by integrating token swap logic. The code will automatically convert the user’s token to ETH before the contract executes the called function.
```

### Governance

> Building bespoke governance systems for a [decentralized autonomous organization](https://ethereum.org/en/dao/) (DAO) can be expensive and time-consuming. However, you can use [Aragon Client](https://client.aragon.org/)—an open-source governance toolkit from Aragon—to bootstrap your DAO and quickly create a usable governance framework. 

Reviewer: - We should explain utility before names (i.e. it should be written that we should be able to drop the name of any project offering this service in place of Aragon, if we wanted)
- Let's assume they know the DAO acronym at this point

```suggestion
Building bespoke governance systems for a [DAO](/dao/) can be expensive and time-consuming. Instead, you could use an open-source governance toolkit, such as [Aragon Client](https://client.aragon.org/), to bootstrap your DAO to quickly create a governance framework. 
```

## What is composability? {#what-is-composability}

Composability is combining distinct components to create new systems or outputs. In software development, composability means developers can reuse existing software components to build new applications. A good way to understand composability is to think of composable elements as Lego blocks. Each Lego can be combined with another, allowing you to build complex structures by combining different Legos. 

In Ethereum, every smart contract is a Lego of sorts—you can use smart contracts from other projects as building blocks for your project. This means you don't have to spend time reinventing the wheel or building from scratch.

## How does composability work? {#how-does-composability-work}

Ethereum smart contracts are like public APIs, so anyone can interact with the contract or integrate them into dapps for added functionality. Smart contract composability generally works off three principles: modularity, autonomy, and discoverability:  

> **1. Modularity**: This is the ability of individual components to perform a specific task. In Ethereum, every smart contract has specific use-case (as shown in the Uniswap example).

Reviewer 2: ```suggestion
**1. Modularity**: This is the ability of individual components to perform a specific task. In Ethereum, every smart contract has a specific use case (as shown in the Uniswap example).
```

### Greater innovation {#greater-innovation}

Composability encourages innovation and experimentation because developers are free to reuse, modify, duplicate, or integrate open-source code to create desired results. As a result, development teams spend less time on basic functionality and can allocate more time experimenting with new features.

### Better user experience {#better-user-experience}

Interoperability between components of the Ethereum ecosystem improves the user experience. Users can access greater functionality when dapps integrate external smart contracts than in a fragmented ecosystem where applications cannot communicate.

## Examples of composability in Ethereum {#composability-in-ethereum}

### Token swaps {#token-swaps}

If you create a dapp that requires transactions to be paid in ETH, you can allow users to pay in other ERC-20 tokens by integrating token swap logic. The code will automatically convert the user’s token to ETH before the contract executes the called function.

### Governance {#governance}

> Building bespoke governance systems for a [DAO](/dao/) can be expensive and time-consuming. Instead, you could use an open-source governance toolkit, such as [Aragon Client](https://client.aragon.org/), to bootstrap your DAO to quickly create a governance framework. 

Reviewer 2: @minimalsm Just noting this product callout. Not necessarily against keeping it since it does demonstrate SC composability, but obviously we've also been discussing trying to not open the flood gates for listings. Curious if you have thoughts.
